File: GVO Code v3/GetQandYU.pas
unit GetQandYU;
//
// Updated for GVO fork mount 6/1/08
//

interface

procedure GetQandY(ra, dec: Real; var Alt, HA, xcount, ycount: Real; var q, y: Integer);

implementation

uses
	Main, Utils, Math;

procedure GetQandY(ra, dec: Real; var Alt, HA, xcount, ycount: Real; var q, y: Integer);

// GetQandY gets quadrant, HA, Alt, xcount, ycount, and ypole of current RA/Dec

begin
	MainForm.Timer_Main.Enabled := False;	// don't allow other calls

	Meridian := GetStime;
	EastHor := Meridian + 6.0;				// ha at east horizon
	WestHor := Meridian - 6.0;				// ha at west horizon

	if (WestHor < 0.0) then
		WestHor := WestHor + 24.0;
	if (EastHor > 24.0) then
		EastHor := EastHor - 24.0;

	EastHA := -6;
	NorthHA := 12;
	WestHA := 6;
	SouthHA := 0;

	// Get Altitude
	HA := 15 * (Meridian - Ra);
	Alt := ArcSin(
		(Sin(DToR * Dec) * Sin(DToR * c_Lat)) +
		(Cos(DToR * Dec) * Cos(DToR * HA) * Cos(DToR * C_Lat))
	);
	Alt := RToD * Alt;	// A = -90 to +90 degrees Altitude

	// Get Ypole
	HA := (Meridian - Ra);
	if (HA < -12) then
		HA := HA + 24;
	if (HA > 12) then
		HA := HA - 24;

	if ((HA <= SouthHA) and (HA >= EastHA)) then
	begin
		Y := 1;
		Q := 1;
	end
	else if ((HA < EastHA) and (HA > -NorthHA)) then
	begin
		Y := -1;
		Q := 2;
	end
	else if ((HA >= SouthHA) and (HA <= WestHA)) then
	begin
		Y := 1;
		Q := 3;
	end
	else if ((HA > WestHA) and (HA < NorthHA)) then
	begin
		Y := -1;
		Q := 4;
	end;

	// Get position counts - Dec first
	// Dec stepper counts positive when north of zenith (> lat),
	// negative when south of zenith (< lat)
	if (Dec >= c_Lat) then
	begin
		// counts positive
		ycount := (Dec - c_Lat) * DecFact;
	end
	else
	begin
		ycount := -1 * ((c_Lat - Dec) * DecFact);
	end;

	// Get RA Counts
	// East of meridian is negative counts (-HA)
	// West of meridian is positive counts (+HA)
	xcount := (HA * 15) * RaFact;

	// From german equatorial model
	// if (Q = 1) then Xcount := ((6.0 - abs(TempHa)) * 15) * RaFact
	// else if (Q = 2) then Xcount := ((abs(TempHa) - 6.0) * 15) * RaFact
	// else if (Q = 3) then Xcount := ((abs(TempHa) - 6.0) * 15) * RaFact
	// else if (Q = 4) then Xcount := ((6.0 - abs(TempHa)) * 15) * RaFact;
	// if ((Q = 4) or (Q = 2)) then Xcount := Xcount * -1.0;

	MainForm.Timer_Main.Enabled := True;
end;

end.

File: GVO Code C++ v1/getQandYU.cpp
#include "GetQandYU.h"
#include "GlobalValues.h"
#include <cmath>

double GetStime();

void GetQandY(double ra, double dec, double& Alt, double& HA, double& Xcount, double& Ycount, int& q, int& y) {
	Meridian = GetStime();
	EastHor = Meridian + 6.0;
	WestHor = Meridian - 6.0;

	if (WestHor < 0.0) WestHor += 24.0;
	if (EastHor > 24.0) EastHor -= 24.0;

	EastHA = -6.0;
	NorthHA = 12.0;
	WestHA = 6.0;
	SouthHA = 0.0;

	// Altitude uses HA in degrees, exactly like Pascal
	double HAdeg = 15.0 * (Meridian - ra);

	Alt = std::asin(
		(std::sin(DToR * dec) * std::sin(DToR * C_Lat)) +
		(std::cos(DToR * dec) * std::cos(DToR * HAdeg) * std::cos(DToR * C_Lat))
	);
	Alt *= RToD;

	// Ypole / quadrant uses HA in hours, exactly like Pascal
	HA = Meridian - ra;

	if (HA < -12.0) HA += 24.0;
	if (HA > 12.0) HA -= 24.0;

	if ((HA <= SouthHA) && (HA >= EastHA)) {
		y = 1;
		q = 1;
	}
	else if ((HA < EastHA) && (HA > -NorthHA)) {
		y = -1;
		q = 2;
	}
	else if ((HA >= SouthHA) && (HA <= WestHA)) {
		y = 1;
		q = 3;
	}
	else if ((HA > WestHA) && (HA < NorthHA)) {
		y = -1;
		q = 4;
	}
	else {
		y = 0;
		q = 0;
	}

	if (dec >= C_Lat) {
		Ycount = (dec - C_Lat) * DECFACT;
	}
	else {
		Ycount = -1.0 * ((C_Lat - dec) * DECFACT);
	}

	Xcount = (HA * 15.0) * RAFact;
}